Webapp/sources/auxiliary.py

```diff
@@ -41,6 +42,161 @@ def smiles_symbols(
+    open_index = -1
+    close_index = -1
+    balance = 0
+
+    # Traverse the expression in reverse to find the last matching pair of brackets
+    for i in range(len(expression) - 1, -1, -1):
+        if expression[i] == ")":
+            if balance == 0:
+                close_index = i
+            balance += 1
+        elif expression[i] == "(":
+            balance -= 1
+            if balance == 0:
+                open_index = i
+                break
+
+    # If we have found a matching set, return the substring
+    if open_index != -1 and close_index != -1:
+        return expression[open_index + 1 : close_index]
+    return ""
+
+
+def reformat_smiles(smiles):
+    branch = smiles.split(")")[-1]
+    smiles = smiles[: smiles.rfind(branch)]
+    inner = get_last_bracketed_expression(smiles)
+    smiles = smiles.split(inner)[0] + branch + ")" + inner
+
+    return smiles
+
+
+def find_polymer_repeat_unit(
+    compound: str,
+) -> str:
+    """
+    Identify the repeat group within a polymer SMILES string.
+    Must be done before smiles is cleaned.
+    Uses {-} and {+n} to find repeat unit.
+    """
+    start_marker = compound.find("{-}")
+    end_marker = compound.find("{+n}")
+
+    # If both {-} and {+} are found
+    if start_marker != -1 and end_marker != -1:
+        # Reformat if end of SRU is treated as a branch. e.g. *C{-}(CC{+n}*)C where {+n} is within brackets
+        if is_within_brackets(compound, "{+n}"):
+            compound = reformat_smiles(compound)
+            start_marker = compound.find("{-}")
+            end_marker = compound.find("{+n}")
+
+        # Start is the character just before {-}, and end is the character just before {+}
+        result = compound[start_marker - 1] + compound[start_marker + 3 : end_marker]
+
+        # Check for branching: if the character after {+} is '('
+        if end_marker + 4 < len(compound) and compound[end_marker + 4] == "(":
+            branch, close_paren = extract_inside_brackets(compound, end_marker + 4)
+            if close_paren != -1:
+                result += branch
+
+            # check for more branching
+            while close_paren + 1 < len(compound) and compound[close_paren + 1] == "(":
+                branch, close_paren = extract_inside_brackets(compound, close_paren + 1)
+                if close_paren != -1:
+                    result += branch
+        return result
+    # if either {-} or {+} is not found
+    return ""
+
+
+def clean_polymer_smiles(
+    compound: str,
+) -> str:
+    """
+    Removes polymer symbols from the string generated by sketchers. last one removes dangling bonds
+    """
+    return compound.replace("{+n}", "").replace("{-}", "").replace("-", "")
+
+
+def canonicalise(smiles):
+    """
+    Canonicalise a polymer SMILES string. Not for use when endgroups are known (must have two '*'s)
+    """
+    ps = PolymerSmiles(smiles)
+    smiles = ps.canonicalize
+    return str(smiles).replace("[", "").replace("]", "")
+
+
```

Comment:
If these are only used for polymers it might be worth moving them into a srouces.services.polymer.py file where we could keep all polymer-related functions. We're generally trying to move away from the auxiliary.py file